feat: Implement method to get required server block and location block / set default configuration values

S0YKIM opened the pull request:
## 작업 내용
- closes #34
- Get server block by port number from Config instance
- Get location block by uri from Block instance(server block)

## 리뷰어에게
- macro.hpp 에 하나하나 값을 정의하지 않고 default configuration file 을 따로 저장하여 파싱하는 방식으로 변경되었습니다
- 기타 자세한 내용은 커밋 메시지 참고 바랍니다

S0YKIM committed:
feat: Implement method to get server block and location block with required port and uri

Additional Tasks
- Add InvalidConfigFileException
- Add InvalidLocationBlockException

S0YKIM committed:
feat: Add default configuration file (#34)

S0YKIM committed:
feat: Implement methods to get required blocks from Config class (#34)

- Get server block from config instance with port number (eg: conf.getServerBlock(8000))
- Get location block from server block with uri (eg: serverblock.getLocationBlock(*.py))
- Set directive value from default configuration file if there is no required directive in provided configuration file
- Implement nested location blocks

Additional Tasks
- Fix bug which pointer to parent in location block points to garbage value by dynamic allocation
- Change type of directive cgi from set to string
- Change name of a directive from limit_except to allowed_methods
- Move static variable defaultBlock from Config class to Block class and initialize it in Config class constructor

hhkim0729 commented on srcs/config/Block.cpp:
```diff
@@ -171,131 +243,158 @@ void Block::parseLocationBlock(std::vector<std::string> &tokens, int &i) {
 	hostport = parseHostPort(args[0]);
-	host_ = hostport[0];
-	port_ = atoi(hostport[1].c_str());
+	setHost(hostport[0]);
+	setPort(atoi(hostport[1].c_str()));
```
바로 저장을 안하고 setter 호출하는 이유가 있나요?

S0YKIM replied:
직접 저장해도 상관 없는데 보기에 예뻐서 이렇게 했어요!

hhkim0729 replied:
setter 내부에서 가공이 필요 없는 경우에는 함수 호출 시 오버헤드가 늘어나기 때문에 직접 멤버 변수에 할당하는 것이 좋아보입니다~!

S0YKIM replied:
해당 부분 수정하였습니다!

hhkim0729 commented on srcs/config/Config.hpp:
```diff
@@ -4,26 +4,29 @@
 # define BUFFER_SIZE 1600
 
 # include "Block.hpp"
+# include <stdlib.h>
```
이 헤더는 어떤 것 때문에 쓰이는 걸까요?

S0YKIM replied:
default configuration file  을 따로 저장하여 파싱하는 방식을 쓰기 이전에는 macro.hpp 에 모든 값을 정의해두고 써야했었는데, 그때 itoa() 가 필요해서 추가한 헤더 파일이었습니다 지금은 해당 루틴이 삭제되어서 필요 없는 헤더 파일이 되었으니 삭제하도록 하겠습니다~

hhkim0729 reviewed (changes requested):
복잡한 Config 구현하느라 고생하셨습니다~!
LocationBlock을 조회할 때 URI가 정확히 일치할 때만 가능한데, URI를 통째로 넘겨도 잘 찾을 수 있도록 그 부분 수정해주시면 될 것 같습니다.

srngch reviewed (approved):
작지않은 클래스인데 너무 고생 많으셨습니다!!

S0YKIM commented:
> 복잡한 Config 구현하느라 고생하셨습니다~! LocationBlock을 조회할 때 URI가 정확히 일치할 때만 가능한데, URI를 통째로 넘겨도 잘 찾을 수 있도록 그 부분 수정해주시면 될 것 같습니다.

@hhkim0729  upload.py abc.py 등등 어떤 이름으로 uri 가 들어오더라도 *.py 로케이션 블록으로 들어오도록 수정하겠습니다 이슈에 새로 추가했어요~

S0YKIM committed:
feat: Find uri with wildcard applied (#34)

- Apply wildcard when the provided uri is extension
- Add supportedExtensions to private member variables in Block class to apply wildcard

Additional Tasks
- Remove parent in private member variables in Block class
- Use copy constructor instead to initialize every value as same as parent
- Change type of serverblocks and locationblocks vector from std::vector<Block *> to std::vector<Block>
- Add validation process to check child's uri contains parent's uri

hhkim0729 commented on srcs/config/Block.cpp:
```diff
@@ -3,44 +3,20 @@
+	if (uri_ == "")
+		return FT_TRUE;
+	len = uri_.length();
+	if (!strncmp(uri.c_str(), uri_.c_str(), len))
+		return FT_TRUE;
```
string을 c_str한 다음 c 함수를 쓰는 것보다 find()나 compare() 같은 string 내장 함수를 사용하는 것이 나을 것 같습니다

S0YKIM replied:
compare() 로 수정 완료했습니다~

hhkim0729 commented on srcs/config/Block.cpp:
```diff
@@ -78,35 +54,121 @@ std::vector<std::string> Block::parseHostPort(const std::string &arg) {
-	setLocationDirectivesMap();
+Block::Block(const Block &origin) {
+	*this = origin;
+	if (origin.uri_ != "")
+		setLocationDirectivesMap();
```
origin의 uri가 없는 경우가 언제인지 궁금합니다~!

S0YKIM replied:
서버 블록의 경우에는 uri 가 없고, 로케이션 블록인 경우에만 uri 를 가지고 있습니다.
예를 들면
```
location aaa {
    ...
}
```
여기서 aaa 가 uri 가 됩니다~

hhkim0729 commented on srcs/config/Block.cpp:
```diff
@@ -404,7 +445,7 @@ const std::string &Block::getErrorPage(int num) const {
 	it = pages.find(num);
 	if (it == pages.end())
-		return Block::defaultBlock_.getErrorPage(num);
+		throw InvalidConfigFileException("There is no error page set in the configuration file.\n");
```
exception 클래스를 따로 만드신 이유가 궁금합니다~! 만약 이 에러가 발생하면 서버가 어떤식으로 동작해야 할까요?

hhkim0729 replied:
디폴트 블록에 에러페이지가 없으면 빈문자열 리턴하기로 했어요~!

S0YKIM replied:
빈 문자열 반환하는 것으로 변경 완료했습니다~

hhkim0729 commented on srcs/config/default.conf:
```diff
@@ -4,7 +4,7 @@ server {
-	client_max_body_size 16000;
+	client_max_body_size 1048576;
 
 	error_page 400 error_pages/400.html;
```
디폴트 에러페이지 삭제하기로 했던 것 같슴다

S0YKIM replied:
삭제 완료했습니다~

hhkim0729 reviewed (changes requested):
드디어 Config 탈출..?
정말 고생하셨습니다~!
 코멘트 참고해서 수정하신 후에 머지하면 될 것 같습니다.

srngch commented on srcs/config/Block.cpp:
```diff
@@ -415,39 +456,48 @@ const std::string &Block::getUri() const {
-	std::vector<Block *>::iterator	it;
-	Block							ret;
+void Block::applyWildCard(std::string &uri, int &dot) const {
+	int			slash = 0;
+
+	for (int i = dot; i >= 0; i--) {
+		if (uri[i] == '/') {
+			slash = i;
+			break ;
+		}
+	}
+	if (slash == dot - 1)
+		return ;
```
string의 search 같은 메소드로 리팩토링이 가능할 것 같습니다~!

S0YKIM replied:
find_last_of() 메소드로 변경 했습니다~

srngch reviewed (approved):
수고하셨습니다 👏 잘 돌아간다면 우선 진행하고, C++ 스타일로 바꾸는 리팩토링은 추후에 진행해도 될 듯해요!

S0YKIM committed:
fix: Remove default error page setting in default config file

Additional Tasks
- Refactor c style code to c++ style
